File: virtual/GR-06/codigo/lu/parallel_lu_matrix_inversion.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <sys/time.h>
#include <string.h>

#define MAX_N 1024
#define MAX_LINE 8192

typedef struct {
    int col;
    int N;
    double **L;
    double **U;
    double **A_inv;
} ThreadData;

double **alocar_matriz(int N) {
    double **M = malloc(N * sizeof(double *));
    for (int i = 0; i < N; i++)
        M[i] = malloc(N * sizeof(double));
    return M;
}

void liberar_matriz(double **M, int N) {
    for (int i = 0; i < N; i++)
        free(M[i]);
    free(M);
}

int leer_csv(const char *filename, double ***A_out, int *N_out) {
    FILE *fp = fopen(filename, "r");
    if (!fp) {
        perror("Error al abrir archivo CSV");
        return 0;
    }

    char line[MAX_LINE];
    int N = 0;
    int columnas = -1;
    double **A = alocar_matriz(MAX_N);

    while (fgets(line, sizeof(line), fp)) {
        int col = 0;
        char *token = strtok(line, ",\n");
        while (token) {
            A[N][col++] = atof(token);
            token = strtok(NULL, ",\n");
        }
        if (columnas == -1) columnas = col;
        else if (col != columnas) {
            fprintf(stderr, "Error: filas con distinta cantidad de columnas.\n");
            liberar_matriz(A, MAX_N);
            fclose(fp);
            return 0;
        }
        N++;
    }

    fclose(fp);
    if (N != columnas) {
        fprintf(stderr, "Error: matriz no es cuadrada.\n");
        liberar_matriz(A, MAX_N);
        return 0;
    }

    *A_out = A;
    *N_out = N;
    return 1;
}

void lu_decomposition(double **A, double **L, double **U, int N) {
    for (int i = 0; i < N; i++) {
        for (int k = i; k < N; k++) {
            double sum = 0.0;
            for (int j = 0; j < i; j++)
                sum += L[i][j] * U[j][k];
            U[i][k] = A[i][k] - sum;
        }

        for (int k = i; k < N; k++) {
            if (i == k)
                L[i][i] = 1.0;
            else {
                double sum = 0.0;
                for (int j = 0; j < i; j++)
                    sum += L[k][j] * U[j][i];
                L[k][i] = (A[k][i] - sum) / U[i][i];
            }
        }
    }
}

void forward_substitution(double **L, double *b, double *y, int N) {
    for (int i = 0; i < N; i++) {
        y[i] = b[i];
        for (int j = 0; j < i; j++)
            y[i] -= L[i][j] * y[j];
    }
}

void backward_substitution(double **U, double *y, double *x, int N) {
    for (int i = N - 1; i >= 0; i--) {
        x[i] = y[i];
        for (int j = i + 1; j < N; j++)
            x[i] -= U[i][j] * x[j];
        x[i] /= U[i][i];
    }
}

void *invert_column(void *arg) {
    ThreadData *data = (ThreadData *)arg;
    int i = data->col, N = data->N;
    double *e = calloc(N, sizeof(double));
    double *y = malloc(N * sizeof(double));
    double *x = malloc(N * sizeof(double));

    e[i] = 1.0;
    forward_substitution(data->L, e, y, N);
    backward_substitution(data->U, y, x, N);

    for (int j = 0; j < N; j++)
        data->A_inv[j][i] = x[j];

    free(e); free(y); free(x);
    return NULL;
}

void invertir_matriz(double **A, double **A_inv, int N) {
    double **L = alocar_matriz(N);
    double **U = alocar_matriz(N);
    lu_decomposition(A, L, U, N);

    pthread_t *threads = malloc(N * sizeof(pthread_t));
    ThreadData *data = malloc(N * sizeof(ThreadData));

    for (int i = 0; i < N; i++) {
        data[i].col = i;
        data[i].N = N;
        data[i].L = L;
        data[i].U = U;
        data[i].A_inv = A_inv;
        pthread_create(&threads[i], NULL, invert_column, &data[i]);
    }

    for (int i = 0; i < N; i++)
        pthread_join(threads[i], NULL);

    liberar_matriz(L, N);
    liberar_matriz(U, N);
    free(threads);
    free(data);
}

void imprimir_matriz(double **M, int N) {
    for (int i = 0; i < N; i++) {
        for (int j = 0; j < N; j++)
            printf("%10.6f ", M[i][j]);
        printf("\n");
    }
}

int main() {
    double **A, **A_inv;
    int N;

    if (!leer_csv("matriz.csv", &A, &N)) {
        fprintf(stderr, "Error leyendo matriz desde CSV.\n");
        return 1;
    }

    A_inv = alocar_matriz(N);

    struct timeval start, end;
    gettimeofday(&start, NULL);

    invertir_matriz(A, A_inv, N);

    gettimeofday(&end, NULL);

    double elapsed = (end.tv_sec - start.tv_sec) + (end.tv_usec - start.tv_usec) / 1e6;

    printf("Matriz inversa A^-1:\n");
    imprimir_matriz(A_inv, N);

    printf("\nTiempo de ejecución: %.6f segundos\n", elapsed);

    liberar_matriz(A, N);
    liberar_matriz(A_inv, N);

    return 0;
}

File: virtual/GR-06/codigo/lu/parallel_lu_matrix_inversion.py
import numpy as np
import threading
import time
import sys

def leer_csv(nombre_archivo):
    try:
        A = np.loadtxt(nombre_archivo, delimiter=",")
        if A.shape[0] != A.shape[1]:
            raise ValueError("La matriz no es cuadrada")
        return A
    except Exception as e:
        print(f"Error al leer el archivo CSV: {e}")
        return None

def lu_decomposition(A):
    N = A.shape[0]
    L = np.zeros((N, N))
    U = np.zeros((N, N))

    for i in range(N):
        for k in range(i, N):
            suma = sum(L[i][j] * U[j][k] for j in range(i))
            U[i][k] = A[i][k] - suma

        for k in range(i, N):
            if i == k:
                L[i][i] = 1.0
            else:
                suma = sum(L[k][j] * U[j][i] for j in range(i))
                L[k][i] = (A[k][i] - suma) / U[i][i]

    return L, U

def forward_substitution(L, b):
    N = L.shape[0]
    y = np.zeros(N)
    for i in range(N):
        y[i] = b[i] - np.dot(L[i, :i], y[:i])
    return y

def backward_substitution(U, y):
    N = U.shape[0]
    x = np.zeros(N)
    for i in reversed(range(N)):
        x[i] = (y[i] - np.dot(U[i, i+1:], x[i+1:])) / U[i][i]
    return x

def invertir_columna(i, L, U, A_inv, N):
    e = np.zeros(N)
    e[i] = 1.0
    y = forward_substitution(L, e)
    x = backward_substitution(U, y)
    A_inv[:, i] = x

def invertir_matriz(A):
    N = A.shape[0]
    L, U = lu_decomposition(A)
    A_inv = np.zeros_like(A)

    hilos = []
    for i in range(N):
        hilo = threading.Thread(target=invertir_columna, args=(i, L, U, A_inv, N))
        hilos.append(hilo)
        hilo.start()

    for hilo in hilos:
        hilo.join()

    return A_inv

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Uso: python3 parallel_lu_matrix_inversion.py <dimensión>")
        sys.exit(1)

    try:
        N = int(sys.argv[1])
        if N <= 0:
            raise ValueError()
    except ValueError:
        print("Error: La dimensión debe ser un entero positivo")
        sys.exit(1)

    archivo = f"../../matrices/matriz_{N}x{N}_invertible.csv"
    A = leer_csv(archivo)

    if A is not None:
        start = time.time()
        A_inv = invertir_matriz(A)
        end = time.time()

        if N <= 10:
            print("\nMatriz inversa A^-1:")
            print(np.round(A_inv, 6))
        else:
            np.savetxt("inversa.csv", A_inv, delimiter=",", fmt="%.10f")
            print("\nMatriz inversa guardada en 'inversa.csv' (no se muestra por ser mayor de 10x10).")

        print(f"\nTiempo de ejecución: {end - start:.6f} segundos")
